File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-7.cpp
class Solution {
public:
    int findMin(vector<int> &nums) {
        int l = 0, r = nums.size() - 1;
        if (nums[0] < nums[1] && nums[0] < nums[nums.size()-1]) return nums[0];
        while (l < r) {
            int middle = static_cast<int> ((l + r) / 2);
            if ((middle != 0 && nums[middle] < nums[middle-1])) return nums[middle];
            // case middle > r and l
            else if (nums[middle] > nums[r]) l = middle + 1;
            else r = middle - 1; 
        }
        return nums[l];
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-7.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        if head is None or head.next is None: return
        # find middle and reverse
        fast = slow = head
        prevslow = None
        while fast and fast.next:
            fast = fast.next.next
            prevslow = slow
            slow = slow.next
        # slow is now our middle
        prevslow.next = None
        prev = None
        nnode = slow
        while slow:
            nnode = nnode.next
            slow.next = prev
            prev = slow
            slow = nnode
        first, second = head, prev


        while first and second:
            nnode = first.next
            first.next = second
            first = nnode
            if not first and second.next != None: break
            nnode = second.next
            second.next = first
            second = nnode
        


        
